test/test_hpytype.py

```diff
@@ -16,51 +16,42 @@ class PointTemplate(DefaultExtensionTemplate):
     of markers, to test different features.
     """
 
-    LEGACY = False
+    _CURRENT_STRUCT = None
+
+    _STRUCT_BEGIN_FORMAT = """
+        typedef struct {{
+    """
+
+    _STRUCT_END_FORMAT = """
+        }} {struct_name};
+        HPyType_HELPERS({struct_name})
+    """
+
+    _IS_LEGACY = "/* not a legacy type */"
+
+    def TYPE_STRUCT_BEGIN(self, struct_name):
+        assert self._CURRENT_STRUCT is None
+        self._CURRENT_STRUCT = struct_name
+        return self._STRUCT_BEGIN_FORMAT.format(struct_name=struct_name)
+
+    def TYPE_STRUCT_END(self):
+        assert self._CURRENT_STRUCT is not None
+        struct_name = self._CURRENT_STRUCT
+        self._CURRENT_STRUCT = None
+        return self._STRUCT_END_FORMAT.format(struct_name=struct_name)
```

Comment:
sorry for being pedantic, but this looks ugly to me :(
Things like these are very hard to understand without manually "parsing and formatting" the code in your head, especially if it's the first time you read the file:
```c
            @TYPE_STRUCT_BEGIN(FooObject)
                float FLOAT_member;
                double DOUBLE_member;
                const char* STRING_member;
                char CHAR_member;
                char ISTRING_member[6];
                char BOOL_member;
            @TYPE_STRUCT_END
```

If we **really** want to go in this direction, then it means that it's probably time to switch to a proper templating engine instead of our custom one.
However, I am skeptical that this is the way to go, because as soon as you start putting too much logic in the template you quickly make your test unreadable.
What about the following:
```c
typedef struct {
    @MAYBE_PyObject_HEAD
    long x;
    long y;
} PointObject;
@HPyType_HELPERS(PointObject);
```

Reply:
Your suggestion is a bit like how things were before, except that we will also need `@MAYBE_Python_H`, and then people have to remember to put three strange things in different places. With `TYPE_STRUCT_BEGIN` and `TYPE_STRUCT_END`, at least there is a clear way to define a type in these tests and it either fails straight away or does the right thing if you use it.

Reply:
@antocuni Once the tests pass, I'll merge this and if we come up with a better plan for structuring the tests I can make a new PR for it.